VGV952CJW33-E-IR: mtd: nand: samsung: Disable subpage writes on E-die NAND

Quallenauge opened the pull request:
Samsung E-die SLC NAND manufactured using 21nm process (K9F1G08U0E)
does not support partial page programming, so disable subpage writes
for it. Manufacturing process is stored in lowest two bits of 5th ID
byte.

This also applies also on some Samsung NANDs (there are different
nands used) which are used on the Easybox 904xDSL device.

Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Please remove this message before posting the pull request.

Quallenauge commented:
@arnysch It would be cool if you can try this and add an printk statement which shows if your NAND is also detected as "non-subpage-writeable" (which would be wrong) or if it is detected correctly.

arnysch commented:
Correctly detected here as [sub-page-writable](http://arny.tjps.eu/OpenWrt/EasyBox904xDSL/U-Boot/NAND-Subpage-Log-for-Qauge.txt).

Quallenauge committed:
VGV952CJW33-E-IR: mtd: nand: samsung: Disable subpage writes on 21nm NAND.

Some Samsung SLC NAND are manufactured using the 21nm process.
They does not supports partial page programming, so disable subpage writes
for it. Manufacturing process is stored in lowest two bits of 5th ID
byte.

This patch is derived and adapted from the upstream patch which
handles a different samsung NAND flash devie (K9F1G08U0E) and is named
mtd: nand: samsung: Disable subpage writes on E-die NAND

Quallenauge commented:
@arnysch I uploaded a new patch which removes the ecc property settings (which doesn't seem to affect our device).

arnysch commented:
Related to these ecc property setting: They might be present in 4.19 kernels for good reasons, maybe they are relevant for other devices. Or they might be nonsense. I don't know and I have not studied the code and tried to understand it.
On the other hand the current nand_samsung.c patch is for 4.14 which does not yet have this chip specific ecc stuff, so it is ok if we only add the details which are relevant for us.

Quallenauge commented:
I left that out, because there's no need for our device and I think therefore it's not necessary do include this stuff. Once the kernel is updated in future OpenWRT versions it will cause more headaches when merging this. So I try this today evening and merge it afterwards.

arnysch commented:
1. Hope you added these debug lines because you thought this is a good idea, not because of me.
Because my intention only was to show what I did to produce that specific log.
2. This comparison "if (chip->id.len >= 8)":
Guess for eb904, len is always 6, so following debug output statement will not be executed.
Is this intended?
BTW, the id.data[] array has 8 bytes, so I thought it won't hurt to show all of them, even the unused ones.

Quallenauge commented:
I like the debug statements. They enable the possibility to check if something goes wrong. And with pr_debug they won't hurt performance in any unreasonable way. So thanks for the pass - it's really helpful! :+1: 
I pushed another PR and dropped the length check. Hope it is ok now.

arnysch reviewed (approved):
Looks ok.
Tested. Works for me :-)

Quallenauge commented:
@arnysch Thanks for taking time to test this and give feedback! :100:  * :+1:
